File: FindFirstAndLastElement.cpp
// Time Complexity : O(logN) where N=Number of Elements in array
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
//Implemented using Binary Search approach.

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {

        vector<int> res;
        if(nums.size() == 0)
        {
            res.push_back(-1);
            res.push_back(-1);
            return res;
        }
        int start = BinSerachStart(nums,target);
        int end = BinSerachEnd(nums,target);
        res.push_back(start);
        res.push_back(end);
        return res;
    }

    int BinSerachStart(vector<int> &nums, int target)
    {
        int start = 0,end= nums.size()-1;
        while(start<=end)
        {
            int mid = start + (end-start)/2;
            if(nums[mid] == target)
            {
                if(mid == 0 || nums[mid]>nums[mid-1])
                {
                    return mid;
                }
                else
                    end = mid -1;
            }
            else if(nums[mid]<target)
            {
                start = mid + 1;
            }
            else
            {
                end = mid - 1;
            }
        }
        return -1;
    }
    int BinSerachEnd(vector<int> &nums, int target)
    {
        int start = 0,end= nums.size()-1;
        while(start<=end)
        {
            int mid = start + (end-start)/2;
            if(nums[mid] == target)
            {
                if(mid == nums.size()-1 || nums[mid]<nums[mid+1])
                {
                    return mid;
                }
                else
                    start = mid +1;
            }
            else if(nums[mid]<target)
            {
                start = mid + 1;
            }
            else
            {
                end = mid - 1;
            }
        }
        return -1;
    }
};

File: src/findFirstandLastElement.java
// Time Complexity: log(n);
// Space Complexity: o(1);
// Did it run on leetcode: Yes
// Problems faced: No

public class findFirstandLastElement {

    public int[] searchRange(int[] nums, int target) {

        if(nums == null || nums.length ==0 ){
            return new int[]{-1,-1};
        }


        if( nums[0] > target || nums[nums.length-1] < target ){
            return new int[]{-1,-1};
        }

        int first = helper1(nums,0, nums.length-1, target);

        if( first == -1){
            return new int[]{-1,-1};
        }

        int last = helper2(nums,first, nums.length-1, target);

        return new int[]{first,last};

    }

    public int helper1(int[] nums, int low, int high, int target){

        while(low<=high){

            int mid= low + (high-low)/2;
            if( nums[mid] == target){

                if(mid ==0 || nums[mid] > nums[mid-1]){
                    return mid;
                }else{
                    high = mid-1;
                }

            }else if( nums[mid] > target){
                high = mid-1;
            }else{
                low = mid+1;
            }

        }

        return -1;
    }

    public int helper2(int[] nums, int low, int high, int target){

        while(low<=high){

            int mid= low + (high-low)/2;
            if( nums[mid] == target){

                if(mid ==nums.length-1 || nums[mid] < nums[mid+1]){
                    return mid;
                }else{
                    low = mid+1;
                }

            }else if( nums[mid] > target){
                high = mid-1;
            }else{
                low = mid+1;
            }

        }

        return -1;
    }


}
